### PR TITLE
GPU-accelerate vppnl_nuc_grad for GTH pseudopotentials

### PR DESCRIPTION
**Pull request by Efithor**
Add CUDA kernels for int1e_r2_origi, int1e_r4_origi and ip2 derivatives in overlap.cu. GPU-accelerate vppnl_nuc_grad via merged-Cell cross-basis integrals and cupy contraction.

Tests added for C, Si, Fe (cross-basis integrals, full gradient vs CPU,
finite difference).

### CODE REVIEW

File: gpu4pyscf/pbc/grad/tests/test_pbc_pp_grad.py [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+            else:
+                e_minus = e_d
+
+        fd_grad = (e_plus - e_minus) / (2 * disp)
+        self.assertAlmostEqual(
+            grad[atom_id, cart_id], fd_grad, places,
+            f"FD check: analytical={grad[atom_id,cart_id]:.6e}, fd={fd_grad:.6e}")
+
+    def test_carbon_fd(self):
+        self._fd_check(cell_c, atom_id=1, cart_id=0, places=2)
```
Comment (sunqm) [inline gpu4pyscf/pbc/grad/tests/test_pbc_pp_grad.py:239]: Can this get better agreement with the finite difference checker?

File: gpu4pyscf/pbc/grad/pp.py [hunk 2]
```diff
@@ -17,75 +17,89 @@
...
 from pyscf.pbc.lib.kpts_helper import gamma_point
-from pyscf.pbc.gto.pseudo.pp_int import fake_cell_vnl, _int_vnl, _contract_ppnl_nuc_grad
+from pyscf.pbc.gto.pseudo.pp_int import fake_cell_vnl
 
-# The following function is copied from pyscf/pbc/gto/pseudo/pp_int.py
+# The following function is derived from pyscf/pbc/gto/pseudo/pp_int.py.
 # It's updated to support k-point sampling after pyscf>2.11.0,
-# however we want gpu4pyscf to be compatable with older version of pyscf,
+# however we want gpu4pyscf to be compatible with older versions of pyscf,
 # particularly pyscf==2.8.0, the version used by github CI.
```
Comment (henryw7) [inline gpu4pyscf/pbc/grad/pp.py]: You can remove these old comments, as a gpu version of the pp gradient is implemented.

**Comment by henryw7** [comment]
Why not also replace the non-local piece of pp in Fock calculation, at https://github.com/pyscf/gpu4pyscf/blob/ad09bb0fa9c25cbcbeebfb86cd3f2e4603fab00f/gpu4pyscf/pbc/dft/multigrid.py#L948
https://github.com/pyscf/gpu4pyscf/blob/ad09bb0fa9c25cbcbeebfb86cd3f2e4603fab00f/gpu4pyscf/pbc/dft/multigrid_v2.py#L1153
https://github.com/pyscf/gpu4pyscf/blob/ad09bb0fa9c25cbcbeebfb86cd3f2e4603fab00f/gpu4pyscf/pbc/df/rsdf_builder.py#L932
https://github.com/pyscf/gpu4pyscf/blob/ad09bb0fa9c25cbcbeebfb86cd3f2e4603fab00f/gpu4pyscf/pbc/df/aft.py#L102

File: gpu4pyscf/pbc/grad/tests/test_pbc_pp_grad.py [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+            coords_d[atom_id, cart_id] += sign * disp
+            atom_list = []
+            for ia in range(cell.natm):
+                sym = cell.atom_symbol(ia)
+                atom_list.append([sym, coords_d[ia]])
+            cell_d.atom = atom_list
+            cell_d.build()
+            mf_d = scf.RHF(cell_d)
+            mf_d.max_cycle = 3
+            mf_d.conv_tol = 1e-8
```
Comment (henryw7) [inline gpu4pyscf/pbc/grad/tests/test_pbc_pp_grad.py]: For a finite difference calculation, I usually make the SCF convergence as tight as possible, like 1e-12. I think this can help you get more digits for comparsion. Also I think a SCF for each FD step is completely unnecessary, you can use a fixed random density matrix at all FD steps, and contract it with the displaced vpp Fock matrix to get vpp energy at each FD step.

File: gpu4pyscf/pbc/grad/tests/test_pbc_pp_grad.py (hunk 1 above)
Comment (Efithor) [reply gpu4pyscf/pbc/grad/tests/test_pbc_pp_grad.py]: Yep Yep - fixed to use a frozen DM with conv_tol=1e-12. Agreement is now 5 decimal places for C, Si and 4 for Fe.

File: gpu4pyscf/pbc/grad/pp.py (hunk 2 above)
Comment (Efithor) [reply gpu4pyscf/pbc/grad/pp.py]: o7

File: gpu4pyscf/pbc/grad/tests/test_pbc_pp_grad.py (hunk 3 above)
Comment (Efithor) [reply gpu4pyscf/pbc/grad/tests/test_pbc_pp_grad.py]: Updated with conv_tol to 1e-12 via frozen converged DM (TestVppnlNucGrad covers random-DM case)

**Comment by Efithor** [comment]
> Why not also replace the non-local piece of pp in Fock calculation, at
> 
> https://github.com/pyscf/gpu4pyscf/blob/ad09bb0fa9c25cbcbeebfb86cd3f2e4603fab00f/gpu4pyscf/pbc/dft/multigrid.py#L948
> 
> 
> https://github.com/pyscf/gpu4pyscf/blob/ad09bb0fa9c25cbcbeebfb86cd3f2e4603fab00f/gpu4pyscf/pbc/dft/multigrid_v2.py#L1153
> 
> 
> https://github.com/pyscf/gpu4pyscf/blob/ad09bb0fa9c25cbcbeebfb86cd3f2e4603fab00f/gpu4pyscf/pbc/df/rsdf_builder.py#L932
> 
> 
> https://github.com/pyscf/gpu4pyscf/blob/ad09bb0fa9c25cbcbeebfb86cd3f2e4603fab00f/gpu4pyscf/pbc/df/aft.py#L102

@henryw7 - Sure, though - I'd prefer to do so in a follow up PR to keep current scope, if that's alright.

File: gpu4pyscf/pbc/grad/tests/test_pbc_pp_grad.py [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+        self._compare_integrals(cell_fe, places=5)
+
+    def test_iron_deriv(self):
+        self._compare_integrals(
+            cell_fe,
+            ('int1e_ipovlp', 'int1e_r2_origi_ip2', 'int1e_r4_origi_ip2'),
+            comp=3, places=5)
+
+
+class TestVppnlNucGrad(unittest.TestCase):
```
Comment (henryw7) [inline gpu4pyscf/pbc/grad/tests/test_pbc_pp_grad.py:179]: Add some tests with non-zero k-points, and I think it's good to merge.

Comment (Efithor) [reply gpu4pyscf/pbc/grad/tests/test_pbc_pp_grad.py:179]: Added in [7472e15](https://github.com/pyscf/gpu4pyscf/pull/740/commits/7472e159dea2806c2f4bac0197c45976d25a32c5)

**Review by henryw7 — APPROVED** [review]
